File: pkg/mcapencrypt/rotate.go
package mcapencrypt

import (
	"bytes"
	"crypto/ecdh"
	"crypto/rsa"
	"crypto/x509"
	"encoding/binary"
	"encoding/pem"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"time"
)

// RotateKeys reads an encrypted MCAP from r, unwraps the symmetric key using
// oldPrivateKeyPem, re-wraps it for each key in newPublicKeyPems, and writes
// the result to w. Encrypted chunk data is copied verbatim; no message
// decryption occurs. The output is readable by any of the new private keys.
func RotateKeys(r io.Reader, w io.Writer, oldPrivateKeyPem string, newPublicKeyPems []string) error {
	if len(newPublicKeyPems) == 0 {
		return fmt.Errorf("at least one new public key is required")
	}

	// Parse the old private key.
	privKey, err := parsePrivateKeyPEM(oldPrivateKeyPem)
	if err != nil {
		return fmt.Errorf("parse old private key: %w", err)
	}
	var unwrap func(kekAlg string, wrappedKey []byte) ([]byte, error)
	switch k := privKey.(type) {
	case *rsa.PrivateKey:
		unwrap = func(kekAlg string, wrappedKey []byte) ([]byte, error) {
			if kekAlg != "rsa-oaep-sha256" {
				return nil, fmt.Errorf("private key is RSA but slot uses %q", kekAlg)
			}
			return UnwrapSymmetricKey(wrappedKey, k)
		}
	case *ecdh.PrivateKey:
		unwrap = func(kekAlg string, wrappedKey []byte) ([]byte, error) {
			if kekAlg != "x25519-hkdf-xchacha20poly1305" {
				return nil, fmt.Errorf("private key is X25519 but slot uses %q", kekAlg)
			}
			return UnwrapSymmetricKeyX25519(wrappedKey, k)
		}
	default:
		return fmt.Errorf("unsupported private key type %T", privKey)
	}

	// Read the entire input so we can do a single-pass scan.
	raw, err := io.ReadAll(r)
	if err != nil {
		return fmt.Errorf("read input: %w", err)
	}
	if len(raw) < 8 {
		return fmt.Errorf("input too short to be an MCAP file")
	}
	if string(raw[:8]) != mcapMagic {
		return fmt.Errorf("not an MCAP file (bad magic bytes)")
	}
	src := bytes.NewReader(raw[8:])

	// ------------------------------------------------------------------ scan
	type pendingRecord struct {
		opcode byte
		data   []byte
	}
	var schemaRecs []pendingRecord
	var channelRecs []pendingRecord

	type chunkMeta struct {
		dataBufOffset  int64 // byte offset within dataBuf where this chunk record starts
		recordLen      int64 // 9 (header) + len(record data)
		msgStart       uint64
		msgEnd         uint64
		compression    string
		compressedSize uint64
		uncompSize     uint64
	}
	var chunkMetas []chunkMeta

	var symKey []byte
	var fileID []byte
	var wkaCount int

	// headerBuf accumulates records that precede the key attachments in the output:
	//   Header, Schema, Channel records.
	// dataBuf accumulates everything that follows the key attachments:
	//   Metadata, EncryptedChunk, EncryptedAttachment, and other user attachments.
	var headerBuf bytes.Buffer
	var dataBuf bytes.Buffer

	// seenFirstChunk tracks whether we have passed the header section. Records
	// that arrive before any chunk go to headerBuf; after the first chunk they go
	// to dataBuf. In practice Schema/Channel always precede chunks.
	var seenFirstChunk bool

	for {
		opcode, data, readErr := ReadRecord(src)
		if readErr != nil {
			if readErr == io.EOF {
				break
			}
			return fmt.Errorf("read record: %w", readErr)
		}

		switch opcode {
		case opcodeHeader:
			// Always goes to headerBuf (appears before schemas/channels).
			if _, err := headerBuf.Write(makeRotateRecord(opcode, data)); err != nil {
				return err
			}

		case opcodeSchema:
			if _, err := headerBuf.Write(makeRotateRecord(opcode, data)); err != nil {
				return err
			}
			cp := make([]byte, len(data))
			copy(cp, data)
			schemaRecs = append(schemaRecs, pendingRecord{opcode, cp})

		case opcodeChannel:
			if _, err := headerBuf.Write(makeRotateRecord(opcode, data)); err != nil {
				return err
			}
			cp := make([]byte, len(data))
			copy(cp, data)
			channelRecs = append(channelRecs, pendingRecord{opcode, cp})

		case opcodeMetadata:
			if _, err := dataBuf.Write(makeRotateRecord(opcode, data)); err != nil {
				return err
			}

		case opcodeAttach:
			name, mediaType, attData, parseErr := parseAttachmentRecord(data)
			if parseErr != nil {
				return fmt.Errorf("parse attachment: %w", parseErr)
			}
			if name == AttachmentName && mediaType == AttachmentMediaType {
				// Wrapped-key attachment: try to unwrap, then skip (will be replaced).
				wkaCount++
				if symKey == nil {
					wkd, decErr := DecodeWrappedKeyData(attData)
					if decErr == nil {
						candidate, unwrapErr := unwrap(wkd.KEKAlg, wkd.WrappedKey)
						if unwrapErr == nil && len(candidate) == 32 {
							symKey = candidate
							fileID = wkd.FileID
						}
					}
				}
				continue // skip; regenerated below
			}
			if name == ManifestAttachmentName && mediaType == ManifestAttachmentMediaType {
				continue // skip; regenerated below
			}
			// User attachment: keep verbatim.
			if _, err := dataBuf.Write(makeRotateRecord(opcode, data)); err != nil {
				return err
			}

		case OpcodeEncryptedChunk:
			seenFirstChunk = true
			ec, decErr := DecodeEncryptedChunk(data)
			if decErr != nil {
				return fmt.Errorf("decode encrypted chunk header: %w", decErr)
			}
			chunkMetas = append(chunkMetas, chunkMeta{
				dataBufOffset:  int64(dataBuf.Len()), // relative; fixed up in pass-B
				recordLen:      int64(9 + len(data)),
				msgStart:       ec.MessageStartTime,
				msgEnd:         ec.MessageEndTime,
				compression:    ec.Compression,
				compressedSize: uint64(len(ec.EncryptedData)),
				uncompSize:     ec.UncompressedSize,
			})
			if _, err := dataBuf.Write(makeRotateRecord(opcode, data)); err != nil {
				return err
			}

		case OpcodeEncryptedAttachment:
			seenFirstChunk = true
			if _, err := dataBuf.Write(makeRotateRecord(opcode, data)); err != nil {
				return err
			}

		case opcodeDataEnd:
			// Stop scanning at DataEnd; regenerate it at the end.
			goto doneScanning

		case opcodeFooter:
			// Reached summary section; stop.
			goto doneScanning

		default:
			// Unknown/index records: put in the appropriate buffer.
			if !seenFirstChunk {
				if _, err := headerBuf.Write(makeRotateRecord(opcode, data)); err != nil {
					return err
				}
			} else {
				if _, err := dataBuf.Write(makeRotateRecord(opcode, data)); err != nil {
					return err
				}
			}
		}
	}

doneScanning:

	if symKey == nil {
		if wkaCount == 0 {
			return fmt.Errorf("no wrapped key attachment found: is this an encrypted MCAP file?")
		}
		return fmt.Errorf("old private key does not match any of the %d recipient key(s) in this file", wkaCount)
	}
	defer clear(symKey)

	// --- Build new wrapped-key attachments. ---
	now := uint64(time.Now().UnixNano())
	var newKeyRecords [][]byte
	for i, pubPem := range newPublicKeyPems {
		pub, loadErr := parsePublicKeyFromPEM(pubPem)
		if loadErr != nil {
			return fmt.Errorf("load new public key %d: %w", i+1, loadErr)
		}
		fingerprint, fpErr := SPKIFingerprint(pub)
		if fpErr != nil {
			return fmt.Errorf("fingerprint new key %d: %w", i+1, fpErr)
		}
		var wrapped []byte
		var kekAlg string
		var wrapErr error
		switch k := pub.(type) {
		case *rsa.PublicKey:
			kekAlg = "rsa-oaep-sha256"
			wrapped, wrapErr = WrapSymmetricKey(symKey, k)
		case *ecdh.PublicKey:
			kekAlg = "x25519-hkdf-xchacha20poly1305"
			wrapped, wrapErr = WrapSymmetricKeyX25519(symKey, k)
		default:
			return fmt.Errorf("unsupported public key type %T for new recipient %d", pub, i+1)
		}
		if wrapErr != nil {
			return fmt.Errorf("wrap key for new recipient %d: %w", i+1, wrapErr)
		}
		wkd := &WrappedKeyData{
			FileID:     fileID,
			KeyID:      fingerprint,
			Algorithm:  "xchacha20poly1305",
			KEKAlg:     kekAlg,
			WrappedKey: wrapped,
		}
		attBytes := buildAttachmentBytes(now, 0, AttachmentName, AttachmentMediaType, wkd.Encode())
		newKeyRecords = append(newKeyRecords, makeRotateRecord(opcodeAttach, attBytes))
	}

	// --- Build manifest attachment. ---
	chunkCount := uint64(len(chunkMetas))
	mac := ComputeManifestHMAC(symKey, chunkCount, fileID)
	var manifestPayload [manifestPayloadSize]byte
	binary.LittleEndian.PutUint64(manifestPayload[:8], chunkCount)
	copy(manifestPayload[8:], mac)
	manifestAttBytes := buildAttachmentBytes(now, 0, ManifestAttachmentName, ManifestAttachmentMediaType, manifestPayload[:])
	manifestRecord := makeRotateRecord(opcodeAttach, manifestAttBytes)

	// ----------------------------------------------------------------- pass-B
	// Compute the absolute byte offset in the output file for each chunk.
	// Output layout (all sizes known at this point):
	//   8 (magic)
	//   + headerBuf
	//   + sum(newKeyRecords)
	//   + manifestRecord
	//   + dataBuf

	prefixSize := int64(8) + int64(headerBuf.Len())
	for _, b := range newKeyRecords {
		prefixSize += int64(len(b))
	}
	prefixSize += int64(len(manifestRecord))

	for i := range chunkMetas {
		chunkMetas[i].dataBufOffset += prefixSize
	}

	// --- Write output ---
	if err := WriteMagic(w); err != nil {
		return err
	}
	if _, err := w.Write(headerBuf.Bytes()); err != nil {
		return err
	}
	for _, b := range newKeyRecords {
		if _, err := w.Write(b); err != nil {
			return err
		}
	}
	if _, err := w.Write(manifestRecord); err != nil {
		return err
	}
	dataBytes := dataBuf.Bytes()
	if _, err := w.Write(dataBytes); err != nil {
		return err
	}

	// DataEnd record (data_end_data_length field = 4 zero bytes per spec).
	if err := WriteRecord(w, opcodeDataEnd, []byte{0, 0, 0, 0}); err != nil {
		return err
	}

	// ----------------------------------------------------------------- summary
	// summaryStart is the file offset of the first summary record.
	summaryStart := prefixSize + int64(len(dataBytes)) + int64(9+4) // DataEnd record: 9-byte hdr + 4-byte payload

	put16 := func(b []byte, v uint16) { binary.LittleEndian.PutUint16(b, v) }
	put32 := func(b []byte, v uint32) { binary.LittleEndian.PutUint32(b, v) }
	put64 := func(b []byte, v uint64) { binary.LittleEndian.PutUint64(b, v) }
	putStr := func(s string) []byte {
		b := make([]byte, 4+len(s))
		binary.LittleEndian.PutUint32(b, uint32(len(s)))
		copy(b[4:], s)
		return b
	}

	type group struct {
		opcode    byte
		absStart  int64
		absLength int64
	}
	var groups []group
	var sumBuf []byte
	written := int64(0)
	emitSumRec := func(op byte, d []byte) {
		hdr := [9]byte{op}
		binary.LittleEndian.PutUint64(hdr[1:], uint64(len(d)))
		sumBuf = append(sumBuf, hdr[:]...)
		sumBuf = append(sumBuf, d...)
		written += int64(9 + len(d))
	}

	// Schema group
	schemaGroupStart := summaryStart + written
	for _, r := range schemaRecs {
		emitSumRec(r.opcode, r.data)
	}
	if l := summaryStart + written - schemaGroupStart; l > 0 {
		groups = append(groups, group{opcodeSchema, schemaGroupStart, l})
	}

	// Channel group
	channelGroupStart := summaryStart + written
	for _, r := range channelRecs {
		emitSumRec(r.opcode, r.data)
	}
	if l := summaryStart + written - channelGroupStart; l > 0 {
		groups = append(groups, group{opcodeChannel, channelGroupStart, l})
	}

	// Statistics
	statsGroupStart := summaryStart + written
	var globalMsgStart, globalMsgEnd uint64
	if len(chunkMetas) > 0 {
		globalMsgStart = chunkMetas[0].msgStart
		globalMsgEnd = chunkMetas[0].msgEnd
		for _, m := range chunkMetas[1:] {
			if m.msgStart < globalMsgStart {
				globalMsgStart = m.msgStart
			}
			if m.msgEnd > globalMsgEnd {
				globalMsgEnd = m.msgEnd
			}
		}
	}
	statsBuf := make([]byte, 8+2+4+4+4+4+8+8+4)
	o := 0
	put64(statsBuf[o:], 0)
	o += 8 // message_count (unknown for encrypted files)
	put16(statsBuf[o:], uint16(len(schemaRecs)))
	o += 2
	put32(statsBuf[o:], uint32(len(channelRecs)))
	o += 4
	put32(statsBuf[o:], 0)
	o += 4 // attachment_count
	put32(statsBuf[o:], 0)
	o += 4 // metadata_count
	put32(statsBuf[o:], uint32(len(chunkMetas)))
	o += 4
	put64(statsBuf[o:], globalMsgStart)
	o += 8
	put64(statsBuf[o:], globalMsgEnd)
	o += 8
	put32(statsBuf[o:], 0)
	o += 4 // channel_message_counts: empty
	emitSumRec(opcodeStatistics, statsBuf[:o])
	groups = append(groups, group{opcodeStatistics, statsGroupStart, summaryStart + written - statsGroupStart})

	// ChunkIndex
	chunkIdxGroupStart := summaryStart + written
	for _, m := range chunkMetas {
		comp := m.compression
		ci := make([]byte, 8+8+8+8+4+8+4+len(comp)+8+8)
		oi := 0
		put64(ci[oi:], m.msgStart)
		oi += 8
		put64(ci[oi:], m.msgEnd)
		oi += 8
		put64(ci[oi:], uint64(m.dataBufOffset))
		oi += 8
		put64(ci[oi:], uint64(m.recordLen))
		oi += 8
		put32(ci[oi:], 0)
		oi += 4 // message_index_offsets: empty
		put64(ci[oi:], 0)
		oi += 8 // message_index_length: 0
		copy(ci[oi:], putStr(comp))
		oi += 4 + len(comp)
		put64(ci[oi:], m.compressedSize)
		oi += 8
		put64(ci[oi:], m.uncompSize)
		oi += 8
		emitSumRec(opcodeChunkIndex, ci[:oi])
	}
	if l := summaryStart + written - chunkIdxGroupStart; l > 0 {
		groups = append(groups, group{opcodeChunkIndex, chunkIdxGroupStart, l})
	}

	// SummaryOffset records
	summaryOffsetStart := summaryStart + written
	for _, g := range groups {
		so := make([]byte, 1+8+8)
		so[0] = g.opcode
		put64(so[1:], uint64(g.absStart))
		put64(so[9:], uint64(g.absLength))
		emitSumRec(opcodeSummaryOffset, so)
	}

	if _, err := w.Write(sumBuf); err != nil {
		return fmt.Errorf("write summary: %w", err)
	}

	// Footer
	footerBuf := make([]byte, 20)
	put64(footerBuf[0:], uint64(summaryStart))
	put64(footerBuf[8:], uint64(summaryOffsetStart))
	// summary_crc = 0 (bytes 16-19 remain zero)
	if err := WriteRecord(w, opcodeFooter, footerBuf); err != nil {
		return err
	}

	return WriteMagic(w)
}

// RotateKeyFile rotates the recipient keys of an encrypted MCAP file.
// It reads inputPath, replaces the wrapped symmetric key with copies wrapped
// for each key in newPublicKeyPems, and writes the result to outputPath.
// The output file is written atomically via a temp file + rename.
// inputPath and outputPath must differ.
func RotateKeyFile(inputPath, outputPath, oldPrivateKeyPem string, newPublicKeyPems []string) (retErr error) {
	absIn, err := filepath.Abs(inputPath)
	if err != nil {
		return fmt.Errorf("resolve input path: %w", err)
	}
	absOut, err := filepath.Abs(outputPath)
	if err != nil {
		return fmt.Errorf("resolve output path: %w", err)
	}
	if absIn == absOut {
		return fmt.Errorf("inputPath and outputPath must differ")
	}
	if _, statErr := os.Stat(absOut); statErr == nil {
		return fmt.Errorf("output file already exists: %q (delete it first)", outputPath)
	}

	inFile, err := os.Open(inputPath)
	if err != nil {
		return fmt.Errorf("open input: %w", err)
	}
	defer inFile.Close()

	tmpFile, err := os.CreateTemp(filepath.Dir(outputPath), ".mcap-rotate-tmp-*")
	if err != nil {
		return fmt.Errorf("create temp output: %w", err)
	}
	tmpPath := tmpFile.Name()
	var tmpClosed bool
	defer func() {
		if !tmpClosed {
			if closeErr := tmpFile.Close(); closeErr != nil && retErr == nil {
				retErr = fmt.Errorf("close temp output: %w", closeErr)
			}
		}
		if retErr != nil {
			os.Remove(tmpPath)
		}
	}()

	if err := RotateKeys(inFile, tmpFile, oldPrivateKeyPem, newPublicKeyPems); err != nil {
		return err
	}
	if err := tmpFile.Close(); err != nil {
		return fmt.Errorf("flush temp output: %w", err)
	}
	tmpClosed = true
	return os.Rename(tmpPath, outputPath)
}

// makeRotateRecord serialises opcode + length + data into a single byte slice.
func makeRotateRecord(opcode byte, data []byte) []byte {
	out := make([]byte, 9+len(data))
	out[0] = opcode
	binary.LittleEndian.PutUint64(out[1:], uint64(len(data)))
	copy(out[9:], data)
	return out
}

// parsePublicKeyFromPEM parses an RSA or X25519 public key from a PEM string.
func parsePublicKeyFromPEM(pemStr string) (any, error) {
	data := []byte(pemStr)
	block, _ := pem.Decode(data)
	if block == nil {
		return nil, fmt.Errorf("no PEM block found in provided key string")
	}
	pub, err := x509.ParsePKIXPublicKey(block.Bytes)
	if err != nil {
		return nil, err
	}
	switch pub.(type) {
	case *rsa.PublicKey, *ecdh.PublicKey:
		return pub, nil
	default:
		return nil, fmt.Errorf("unsupported public key type %T", pub)
	}
}

File: ts/src/rotate.ts
import { BinaryReader, BinaryWriter, safeBigintToNumber } from "./binary.js";
import {
  OP_HEADER,
  OP_SCHEMA,
  OP_CHANNEL,
  OP_ATTACHMENT,
  OP_DATA_END,
  OP_FOOTER,
  OP_METADATA,
  OP_ENCRYPTED_CHUNK,
  OP_ENCRYPTED_ATTACHMENT,
  readMagic,
  readRecord,
  writeRecord,
  writeMagic,
} from "./record.js";
import { decodeEncryptedChunk } from "./chunk.js";
import {
  ATTACHMENT_NAME,
  ATTACHMENT_MEDIA_TYPE,
  MANIFEST_ATTACHMENT_NAME,
  MANIFEST_ATTACHMENT_MEDIA_TYPE,
  MANIFEST_PAYLOAD_SIZE,
  computeManifestHMAC,
  decodeWrappedKeyData,
  encodeWrappedKeyData,
  unwrapSymmetricKey,
  wrapSymmetricKey,
  wrapSymmetricKeyX25519,
  isX25519PublicKeyPem,
  spkiFingerprint,
} from "./key.js";
import { encodeAttachment, parseAttachmentFields } from "./attachment.js";

// Additional opcodes for building the summary section.
const OP_CHUNK_INDEX = 0x08;
const OP_STATISTICS = 0x0a;
const OP_SUMMARY_OFFSET = 0x0e;

type ChunkIndexMeta = {
  msgStart: bigint;
  msgEnd: bigint;
  dataBufOffset: bigint; // byte offset within dataBuf; adjusted to file offset in pass-B
  recordLen: bigint;
  compression: string;
  compressedSize: bigint;
  uncompSize: bigint;
};

/**
 * rotateMcapKeys re-wraps the symmetric key for a new set of recipients without
 * decrypting any chunk data. O(file size) I/O with zero message decryption.
 *
 * @param input          - encrypted MCAP bytes
 * @param oldPrivateKeyPem - PEM of the private key that can currently decrypt
 * @param newPublicKeyPems - PEM(s) of the new recipient public keys
 * @returns              - new encrypted MCAP bytes readable by any newPublicKeyPem
 */
export async function rotateMcapKeys(
  input: Uint8Array,
  oldPrivateKeyPem: string,
  newPublicKeyPems: string | string[],
): Promise<Uint8Array> {
  const pubKeys = Array.isArray(newPublicKeyPems)
    ? newPublicKeyPems
    : [newPublicKeyPems];
  if (pubKeys.length === 0) throw new Error("at least one new public key is required");

  const r = new BinaryReader(input);
  readMagic(r);

  // headerBuf: Header, Schema, Channel records (appear before key attachments in output)
  const headerBuf = new BinaryWriter();
  // dataBuf: everything after key attachments (Metadata, EncryptedChunk, etc.)
  const dataBuf = new BinaryWriter();

  const schemaRecs: Uint8Array[] = [];
  const channelRecs: Uint8Array[] = [];
  const chunkMetas: ChunkIndexMeta[] = [];

  let symKey: Uint8Array | undefined;
  let fileId: Uint8Array | undefined;
  let wkaCount = 0;
  let seenFirstChunk = false;

  // ----------------------------------------------------------------- scan
  scan: while (r.remaining > 0) {
    const rec = readRecord(r);
    if (!rec) break;
    const { opcode, data } = rec;

    switch (opcode) {
      case OP_HEADER:
        writeRecord(headerBuf, opcode, new Uint8Array(data));
        break;

      case OP_SCHEMA: {
        const raw = new Uint8Array(data);
        writeRecord(headerBuf, opcode, raw);
        schemaRecs.push(raw);
        break;
      }

      case OP_CHANNEL: {
        const raw = new Uint8Array(data);
        writeRecord(headerBuf, opcode, raw);
        channelRecs.push(raw);
        break;
      }

      case OP_METADATA:
        writeRecord(dataBuf, opcode, new Uint8Array(data));
        break;

      case OP_ATTACHMENT: {
        const { name, mediaType, attData } = parseAttachmentFields(
          new Uint8Array(data),
        );
        if (name === ATTACHMENT_NAME && mediaType === ATTACHMENT_MEDIA_TYPE) {
          // Wrapped-key attachment: try to unwrap; skip (will be replaced).
          wkaCount++;
          if (!symKey) {
            try {
              const wkd = decodeWrappedKeyData(attData);
              const candidate = await unwrapSymmetricKey(
                wkd.wrappedKey,
                oldPrivateKeyPem,
                wkd.kekAlg,
              );
              if (candidate.length === 32) {
                symKey = candidate;
                fileId = wkd.fileId;
              }
            } catch {
              // wrong key or malformed; try next attachment
            }
          }
          break; // skip; regenerated below
        }
        if (
          name === MANIFEST_ATTACHMENT_NAME &&
          mediaType === MANIFEST_ATTACHMENT_MEDIA_TYPE
        ) {
          break; // skip; regenerated below
        }
        // User attachment: keep verbatim.
        writeRecord(dataBuf, opcode, new Uint8Array(data));
        break;
      }

      case OP_ENCRYPTED_CHUNK: {
        seenFirstChunk = true;
        const ec = decodeEncryptedChunk(new Uint8Array(data));
        chunkMetas.push({
          msgStart: ec.messageStartTime,
          msgEnd: ec.messageEndTime,
          dataBufOffset: BigInt(dataBuf.length), // relative; fixed in pass-B
          recordLen: BigInt(9 + data.length),
          compression: ec.compression,
          compressedSize: BigInt(ec.encryptedData.length),
          uncompSize: ec.uncompressedSize,
        });
        writeRecord(dataBuf, opcode, new Uint8Array(data));
        break;
      }

      case OP_ENCRYPTED_ATTACHMENT:
        seenFirstChunk = true;
        writeRecord(dataBuf, opcode, new Uint8Array(data));
        break;

      case OP_DATA_END:
        break scan;

      case OP_FOOTER:
        break scan;

      default:
        // Unknown records: route to the appropriate buffer.
        if (!seenFirstChunk) {
          writeRecord(headerBuf, opcode, new Uint8Array(data));
        } else {
          writeRecord(dataBuf, opcode, new Uint8Array(data));
        }
        break;
    }
  }

  if (!symKey || !fileId) {
    if (wkaCount === 0) {
      throw new Error("no wrapped key attachment found: is this an encrypted MCAP file?");
    }
    throw new Error(
      `old private key does not match any of the ${wkaCount} recipient key(s) in this file`,
    );
  }

  // ----------------------------------------------------------------- wrap
  const now = BigInt(Date.now()) * 1_000_000n;
  const newKeyRecords: Uint8Array[] = [];
  for (let i = 0; i < pubKeys.length; i++) {
    const pubPem = pubKeys[i]!;
    const keyId = await spkiFingerprint(pubPem);
    const isX25519 = isX25519PublicKeyPem(pubPem);
    const wrappedKey = isX25519
      ? await wrapSymmetricKeyX25519(symKey, pubPem)
      : await wrapSymmetricKey(symKey, pubPem);
    const wkdBytes = encodeWrappedKeyData({
      version: 3,
      fileId,
      keyId,
      algorithm: "xchacha20poly1305",
      kekAlg: isX25519 ? "x25519-hkdf-xchacha20poly1305" : "rsa-oaep-sha256",
      wrappedKey,
    });
    const attBytes = encodeAttachment(
      now, 0n, ATTACHMENT_NAME, ATTACHMENT_MEDIA_TYPE, wkdBytes,
    );
    const rec = new BinaryWriter();
    writeRecord(rec, OP_ATTACHMENT, attBytes);
    newKeyRecords.push(rec.toUint8Array());
  }

  // Manifest
  const chunkCount = BigInt(chunkMetas.length);
  const mac = await computeManifestHMAC(symKey, chunkCount, fileId);
  const manifestPayload = new Uint8Array(MANIFEST_PAYLOAD_SIZE);
  new DataView(manifestPayload.buffer).setBigUint64(0, chunkCount, true);
  manifestPayload.set(mac, 8);
  const manifestAttBytes = encodeAttachment(
    now, 0n, MANIFEST_ATTACHMENT_NAME, MANIFEST_ATTACHMENT_MEDIA_TYPE, manifestPayload,
  );
  const manifestRecordBuf = new BinaryWriter();
  writeRecord(manifestRecordBuf, OP_ATTACHMENT, manifestAttBytes);
  const manifestRecord = manifestRecordBuf.toUint8Array();

  // ----------------------------------------------------------------- pass-B
  // Compute prefix size: magic(8) + headerBuf + newKeyRecords + manifestRecord
  const MAGIC_SIZE = 8n;
  let prefixSize = MAGIC_SIZE + BigInt(headerBuf.length);
  for (const kr of newKeyRecords) prefixSize += BigInt(kr.length);
  prefixSize += BigInt(manifestRecord.length);

  // Fix chunk offsets from dataBuf-relative to absolute file offsets.
  for (const m of chunkMetas) {
    m.dataBufOffset += prefixSize;
  }

  // ----------------------------------------------------------------- assemble
  const out = new BinaryWriter();

  writeMagic(out);
  out.writeBytes(headerBuf.toUint8Array());
  for (const kr of newKeyRecords) out.writeBytes(kr);
  out.writeBytes(manifestRecord);
  const dataBytes = dataBuf.toUint8Array();
  out.writeBytes(dataBytes);

  // DataEnd: 4 zero bytes (data_end_data_length per spec).
  writeRecord(out, OP_DATA_END, new Uint8Array(4));

  // ----------------------------------------------------------------- summary
  const summaryStart =
    prefixSize + BigInt(dataBytes.length) + 9n + 4n; // DataEnd: 9-byte hdr + 4-byte payload

  const sumBuf = new BinaryWriter();
  let written = 0n;

  const emitSumRec = (opcode: number, data: Uint8Array): void => {
    const hdr = new Uint8Array(9);
    hdr[0] = opcode;
    new DataView(hdr.buffer).setBigUint64(1, BigInt(data.length), true);
    sumBuf.writeBytes(hdr);
    sumBuf.writeBytes(data);
    written += BigInt(9 + data.length);
  };

  type SumGroup = { opcode: number; absStart: bigint; absLength: bigint };
  const groups: SumGroup[] = [];

  // Schema group
  const schemaGroupStart = summaryStart + written;
  for (const raw of schemaRecs) emitSumRec(OP_SCHEMA, raw);
  const schemaGroupLength = summaryStart + written - schemaGroupStart;
  if (schemaGroupLength > 0n) groups.push({ opcode: OP_SCHEMA, absStart: schemaGroupStart, absLength: schemaGroupLength });

  // Channel group
  const channelGroupStart = summaryStart + written;
  for (const raw of channelRecs) emitSumRec(OP_CHANNEL, raw);
  const channelGroupLength = summaryStart + written - channelGroupStart;
  if (channelGroupLength > 0n) groups.push({ opcode: OP_CHANNEL, absStart: channelGroupStart, absLength: channelGroupLength });

  // Statistics
  const statsGroupStart = summaryStart + written;
  let globalMsgStart = 0xffffffffffffffffn;
  let globalMsgEnd = 0n;
  for (const m of chunkMetas) {
    if (m.msgStart < globalMsgStart) globalMsgStart = m.msgStart;
    if (m.msgEnd > globalMsgEnd) globalMsgEnd = m.msgEnd;
  }
  if (chunkMetas.length === 0) {
    globalMsgStart = 0n;
  }
  const statsBuf = new BinaryWriter();
  statsBuf.writeUint64(0n); // message_count (unknown for encrypted files)
  statsBuf.writeUint16(schemaRecs.length);
  statsBuf.writeUint32(channelRecs.length);
  statsBuf.writeUint32(0); // attachment_count
  statsBuf.writeUint32(0); // metadata_count
  statsBuf.writeUint32(chunkMetas.length);
  statsBuf.writeUint64(globalMsgStart);
  statsBuf.writeUint64(globalMsgEnd);
  statsBuf.writeUint32(0); // channel_message_counts: empty map
  emitSumRec(OP_STATISTICS, statsBuf.toUint8Array());
  groups.push({
    opcode: OP_STATISTICS,
    absStart: statsGroupStart,
    absLength: summaryStart + written - statsGroupStart,
  });

  // ChunkIndex
  const chunkIdxGroupStart = summaryStart + written;
  for (const m of chunkMetas) {
    const ci = new BinaryWriter();
    ci.writeUint64(m.msgStart);
    ci.writeUint64(m.msgEnd);
    ci.writeUint64(m.dataBufOffset);
    ci.writeUint64(m.recordLen);
    ci.writeUint32(0); // message_index_offsets: empty
    ci.writeUint64(0n); // message_index_length: 0
    ci.writeString(m.compression);
    ci.writeUint64(m.compressedSize);
    ci.writeUint64(m.uncompSize);
    emitSumRec(OP_CHUNK_INDEX, ci.toUint8Array());
  }
  const chunkIdxGroupLength = summaryStart + written - chunkIdxGroupStart;
  if (chunkIdxGroupLength > 0n) {
    groups.push({
      opcode: OP_CHUNK_INDEX,
      absStart: chunkIdxGroupStart,
      absLength: chunkIdxGroupLength,
    });
  }

  // SummaryOffset records
  const summaryOffsetStart = summaryStart + written;
  for (const g of groups) {
    const so = new BinaryWriter();
    so.writeUint8(g.opcode);
    so.writeUint64(g.absStart);
    so.writeUint64(g.absLength);
    emitSumRec(OP_SUMMARY_OFFSET, so.toUint8Array());
  }

  out.writeBytes(sumBuf.toUint8Array());

  // Footer
  const footerBuf = new BinaryWriter();
  footerBuf.writeUint64(summaryStart);
  footerBuf.writeUint64(summaryOffsetStart);
  footerBuf.writeUint32(0); // summary_crc
  writeRecord(out, OP_FOOTER, footerBuf.toUint8Array());

  // Trailing magic
  writeMagic(out);

  return out.toUint8Array();
}
